Build on NixOS

I am trying to build Servo on NixOS, with Nix version 1.11.2. I installed the following packages — `bzip2, cmake, freeglut, freetype, glib, gperf, mesa, which, xorg.libX11, xorg.libXcursor, xorg.xrandr` — although these might not be all dependencies. Anyway, 

```
Error running mach:

    ['build', '--dev']

The error occurred in code that was called by the mach command. This is either
a bug in the called code itself or in the way that mach is calling it.

You should consider filing a bug for this issue.

If filing a bug, please include the full output of mach, including this error
message.

The details of the failure are as follows:

OSError: [Errno 2] No such file or directory

  File "/home/sindikat/progs/servo/python/servo/build_commands.py", line 235, in build
    env=env, cwd=self.servo_crate(), verbose=verbose)
  File "./python/servo/command_base.py", line 69, in call
    return subprocess.call(*args, shell=sys.platform == 'win32', **kwargs)
  File "/nix/store/awbf3jh262m1k16hxgbgngdmxv66gd79-python-2.7.11/lib/python2.7/subprocess.py", line 522, in call
    return Popen(*popenargs, **kwargs).wait()
  File "/nix/store/awbf3jh262m1k16hxgbgngdmxv66gd79-python-2.7.11/lib/python2.7/subprocess.py", line 710, in __init__
    errread, errwrite)
  File "/nix/store/awbf3jh262m1k16hxgbgngdmxv66gd79-python-2.7.11/lib/python2.7/subprocess.py", line 1335, in _execute_child
    raise child_exception
```

I have no idea, how to continue from that.

Ok, I ran the script and also installed some other packages just in case, and now `nix-shell --run './mach build --dev'` barfs at installing `harfbuzz-sys`:

    make[1]: *** [/home/sindikat/progs/servo/.cargo/registry/src/github.com-88ac128001ac3a9a/harfbuzz-sys-0.1.2/harfbuzz/config.h.in] Error 127
    make: *** [all] Error 2
    thread '<main>' panicked at 'assertion failed: Command::new("make").args(&["-R", "-f", "makefile.cargo",
                                &format!("-j{}" , env:: var ( "NUM_JOBS" ) .
                                         unwrap (
                                         ))]).status().unwrap().success()', /home/sindikat/progs/servo/.cargo/registry/src/github.com-88ac128001ac3a9a/harfbuzz-sys-0.1.2/build.rs:14
    note: Run with `RUST_BACKTRACE=1` for a backtrace.

Anyway, thanks for your time and effort :D

--- a/README.md
+++ b/README.md
@@ -177,6 +177,14 @@ With the following environment variable set:
 export LIBCLANG_PATH=$(llvm-config --prefix)/lib64
 ```
 
+#### On nixOS Linux
+
+```sh
+nix-shell etc/shell.nix
+```
+
+You will need to run this in every shell before running mach.
+
 #### On Windows (MSVC)
 
 1. Install Python 3.9 for Windows (https://www.python.org/downloads/release/python-392/). The Windows x86-64 MSI installer is fine. This is required in order to build the JavaScript engine, SpiderMonkey.

--- a/etc/shell.nix
+++ b/etc/shell.nix
@@ -1,0 +1,40 @@
+# This provides a shell with all the necesarry packages required to run mach and build servo
+# NOTE: This does not work offline or for nix-build
+
+with import <nixpkgs> {};
+
+clangStdenv.mkDerivation rec {
+  name = "servo-env";
+
+  buildInputs = [
+    # Native dependencies
+    fontconfig freetype openssl libunwind
+    xlibs.libxcb x11
+
+    gst_all_1.gstreamer
+    gst_all_1.gst-plugins-base
+    gst_all_1.gst-plugins-bad
+
+    rustup
+
+    # Build utilities
+    cmake dbus gcc git pkgconfig which llvm autoconf213 perl yasm m4
+    (python3.withPackages (ps: with ps; [virtualenv pip dbus]))
+  ];
+
+  LIBCLANG_PATH = llvmPackages.clang-unwrapped.lib + "/lib/";
+
+  # Allow cargo to download crates
+  SSL_CERT_FILE = "${cacert}/etc/ssl/certs/ca-bundle.crt";
+
+  # Enable colored cargo and rustc output
+  TERMINFO = "${ncurses.out}/share/terminfo";
+
+  shellHook = ''
+    # Fix missing libraries errors (those libraries aren't linked against, so we need to dynamically supply them)
+    export LD_LIBRARY_PATH=${lib.makeLibraryPath [ xorg.libXcursor xorg.libXrandr xorg.libXi ]}
+    # Fix invalid option errors during linking
+    # https://github.com/mozilla/nixpkgs-mozilla/commit/c72ff151a3e25f14182569679ed4cd22ef352328
+    unset AS
+  '';
+}

